### DISCUSSION TITLE
Accepting client's certificate in server

### DISCUSSION THREAD

**Discussion by Hari-quartic** [FreeOpcUa/opcua-asyncio#505] [Q&A]
Hello,

Earlier I was using prosys OPCUA simulation and use to trust the client's certificate in prosys.

Now I am trying to write a custom server using OPCUA asyncio. 
Can someone help me how to accept the client's certificate from my custom server?

**Comment by leonardo1717** [comment]
Hi there,

I can probably not help you all the way. The code below is what gets me past login.
After login the server fails. I am using UaExpert for testing. I exported its certificate and imported that into the server for which I have used the example certificate and key from asyncua. I hope it  helps you get started.

`
import sys
sys.path.insert(0, "..")
import asyncio

from asyncua import ua, uamethod, Server
from asyncua.crypto.permission_rules import SimpleRoleRuleset
from asyncua.server.users import UserRole
from asyncua.server.user_managers import CertificateUserManager

import json
import asyncpg

# Global variable
server = None           # OPCUA server object
cfg = None              # configuration JSON data object

async def main():
    global server
    global cfg

    # read config file
    cfg = getConfigurationData('./config.json')
    cfg["devices"] = {}
    for device in cfg["devicelist"]:
        cfg["devices"][device["nodeid"]] = device["opcname"]

    # setup our server
    cert_user_manager = CertificateUserManager()
    await cert_user_manager.add_user('user_certificates/uaExpert.cer', name='UaExpert')
    server = Server(user_manager=cert_user_manager)

    server.set_endpoint("opc.tcp://0.0.0.0:4840/")
    await server.init()

    server.set_security_policy([ua.SecurityPolicyType.Basic256Sha256_SignAndEncrypt], permission_ruleset=SimpleRoleRuleset)

    await server.load_certificate("certificate.pem")
    await server.load_private_key("key.pem")
`

**Comment by leonardo1717** [comment]
I found out that the code works with 1 little change

This line

`    server.set_security_policy([ua.SecurityPolicyType.Basic256Sha256_SignAndEncrypt], permission_ruleset=SimpleRoleRuleset)
`

should be

`    server.set_security_policy([ua.SecurityPolicyType.Basic256Sha256_SignAndEncrypt], permission_ruleset=SimpleRoleRuleset())
`

and then it works
